ruby/object_oriented_programming_basics/managing_ruby_projects.md

```diff
@@ -0,0 +1,239 @@
+- What `bundle init` does and what a `Gemfile` is.
+- How to get the most out of `Ruby LSP` VSCode extension.
+
+### Confusion, convention, convenience
+
+Recall how you've worked with the projects in Foundations: the HTML, CSS and JS all lived in separate files, becoming "one" thanks to linking CSS and JS in the HTML file(s). Being able to divide the type of work across multiple files hopefully organized your work better and prevented confusion. Moving forward you can imagine that your files are only going to get longer and more plentiful in number. Thus, having some kind of structure becomes necessary. Additionally, you might've already realized how useful git tracking changes per-file together with GitHub is by making it a lot easier to track down a bug if you know that your last commit only affected your JS file.
+
+Now, it'd be nice to be able to also divide your Ruby work across multiple files since you're about to set out onto the more complicated OOP projects, where many distinct parts are going to work in unison. Not only is that possible, there are also some conventions around Ruby project structure that aim at making it easier for you and others to work with and navigate your project.
+
+### Making use of multiple files
+
+If you are to split your code across multiple files, you first will need to know how to make sure code from one file can be used in another file. There are two main ways to do that:
+
+- `require_relative` which takes in a string of a relative path, like `require_relative 'lib/bogo_sort'`, is going to look for `lib/bogo_sort.rb` relative to the location of the current file.
+- `require` is trickier - there are two ways of using it: one is providing a relative path which is recognized by using `../` or `./` at the start or an absolute path. This works like `require_relative`. Another way is by providing a file name without the extension, like `require 'csv'` is going to look for a `csv.rb` in the Ruby's `$LOAD_PATH` global variable which by default contains the Ruby standard library. There are other file extensions it might look for, but this is not important at this point. In addition to that, if it doesn't find that file in `$LOAD_PATH` it is going to look through installed gems (more on those later) to see if the file is there.
+
+Both of those approaches are going to execute the file, allowing you to use their contents. If you try to require something for the second time, nothing will happen, and the requires will return `false`.
+
+Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:
```

Comment:
I looked into how Bundler gem itself is set up:
https://github.com/rubygems/rubygems/blob/5cedfc8ad923cbc679049ed69fbb1baf1753617f/bundler/lib/bundler.rb#L3-L15

Here are the tests:
https://github.com/rubygems/rubygems/blob/5cedfc8ad923cbc679049ed69fbb1baf1753617f/bundler/spec/spec_helper.rb#L1-L26

They follow that convention BUT as you can see, the tests `require bundler` as if it were external! https://guides.rubygems.org/patterns/#loading-code suggests both as OK.

I think some of the rationale could be read here: https://packaging.rubystyle.guide/#using-require-relative-from-test-to-lib https://docs.rubocop.org/rubocop-packaging/cops_packaging.html
and this seems to connect well with this comment of yours:
> Iirc, its because require_relative has limitations with where you can execute code. That mostly surfaces when you're running tests. You always have to be in the root directory, if you're in any other directory the require_relative breaks.

but from what I understand, it is exactly `require`'s problem - if you provide `require` a relative path, it is relative to the place the file was called from, while `require_relative` is always relative to the file it is in. Consider my perhaps not-so-wise example of my Chess (many files cut for brevity's sake):

```bash
├── lib
│   ├── board_navigator.rb
│   ├── board.rb
│   ├── coordinate.rb
│   ├── display.rb
│   ├── en_passant_pair.rb
│   ├── fen.rb
│   ├── fen_serializer.rb
│   ├── game_driver.rb
│   └─── game.rb
├── chess.rb
├── main.rb
└── spec
    ├── board_spec.rb
    ├── e2e
    │   └── game_driver_spec.rb
    ├── examples.txt
    ├── fen_serializer_spec.rb
    └── spec_helper.rb
```  

My `chess.rb` is a file that just `require_relatives` every single file in `lib/`. It's a shortcut I picked up from someone so I could just do `require_relative ../chess` for example in my tests, like `board_spec.rb`. With this setup, I can run tests wherever I want to. But if I change `require_relative ./lib/board.rb` in `chess.rb` to plain `require`, I can only run the test in the root of the project, because when `board_spec.rb` is loading `chess.rb`, it then gets to the `require` line and the path resolves differently, the `require` in `chess.rb` resolves from the point of view of `board_spec.rb` and fails. `require_relative`  makes sure it is resolved from the point of view of `chess.rb`.

It might be the case that `require_relative` being for in-project dependencies is going to be part of Rubocop at some point: https://github.com/rubocop/rubocop/issues/8748#issuecomment-698735044

IN SUMMARY, the tests should consider the project as an external dependency and thus use `require`, stuff in `lib` should use `require_relative`. BUT if you use `require bundler` like they do in bundler tests, you'd need to have bundler already installed... Or put differently, have `bundler.rb` somewhere in your `$LOAD_PATH` so in people's projects I guess using `require` with a relative path would be OK.

Reply:
Oh and this actually means I need to change my explanations on `require` and `require_relative`, because they're wrong. Huge apologies for that.

Reply:
Thanks for sharing those links @scheals, I've got myself brushed up on this now 😄. I think I might have originally been getting things confused with the way gems load everything from lib into the LOAD_PATH. 

This all sounds solid, happy to resolve this thread.

Reply:
Just to chime in with another point about all of this:

I told scheals somewhere (maybe on discord?) to avoid mentioning [autoload](https://docs.ruby-lang.org/en/3.1/Kernel.html#method-i-autoload), which is another technique that's used to require code.

It's commonly used in gems to provide a kind of "lazy loading" functionality. Like taking the [noticed gem](https://github.com/excid3/noticed/blob/main/lib/noticed.rb) as an example, code is required with `autoload` such that it isn't actually required until the class or module name is used somewhere in the application code. So if an application using this gem is never delivering notifications via slack, that code is just never loaded at all.

This is good to leave out right? I figure learners at this point don't really get anything out of using it.

Reply:
Thanks @josh, no arguments from me, I think its good to leave out too. There definitely doesn't seem to be anything to from it this early, and would just muddy the water.